### PR TITLE
Issue8944/template tokens space allowed

### LINKED ISSUE #8944: [Bug] In 2.15.0 {{# Field}} X {{/ Field}} Appears to Be Broken (Need to handle '# Field' with space)
###### Reproduction Steps

1. Use version 2.15.0
2. Wrap a card or field with {{# Field}} X {{/ Field}}
3. Attempt to view card
4. Verify that wrapped content does not display

After the 2.15.0 update, my cards that are wrapped this way are showing as empty cards, and the fields I have wrapped are just not showing up. It seems like it is no longer checking to see if the field is populated and instead is always not displaying it.

###### Expected Result

Cards and fields displayed when field is populated.

###### Actual Result

Cards and fields are not displayed.

###### Debug info
Refer to the [support page](https://ankidroid.org/docs/help.html) if you are unsure where to get the "debug info".

###### Research
*Enter an [x] character to confirm the points below:*

- [x] I have read the [support page](https://ankidroid.org/docs/help.html) and am reporting a bug or enhancement request specific to AnkiDroid
- [x] I have checked the [manual](https://ankidroid.org/docs/manual.html) and the [FAQ](https://github.com/ankidroid/Anki-Android/wiki/FAQ) and could not find a solution to my issue
- [x] I have searched for similar existing issues here and on the user forum
- [ ] (Optional) I have confirmed the issue is not resolved in the latest alpha release ([instructions](https://docs.ankidroid.org/manual.html#betaTesting))

### PR DESCRIPTION
**Pull request by mikehardy**
## Purpose / Description

Template tokens can apparently have space between the token type and the token name.

There is a widely used deck that uses this apparently, and AnkiDesktop allows the spacing per Damien

## Fixes
Fixes #8944

## Approach

Add a fresh test to the wonderful existing test suite from @Arthur-Milchior (thank you, thank you) that proves it fails in the conditions laid out from the original issue reporter

Test fails

Add a replaceFirst call before returning the token, where the regex matches any space chars and removes them

Test passes

## How Has This Been Tested?

Unit testing

## Learning (optional, can help others)

Template parsing is hard because there are so many edge cases

## Checklist
_Please, go through these checks before submitting the PR._

- [x] You have not changed whitespace unnecessarily (it makes diffs hard to read)

Does it count if I automated the changing of whitespace on purpose :thinking: 

- [x] You have a descriptive commit message with a short title (first line, max 50 chars).
- [x] Your code follows the style of the project (e.g. never omit braces in `if` statements) 
- [x] You have commented your code, particularly in hard-to-understand areas
- [x] You have performed a self-review of your own code
- [ ] UI changes: include screenshots of all affected screens (in particular showing any new or changed strings)
- [ ] UI Changes: You have tested your change using the [Google Accessibility Scanner](https://play.google.com/store/apps/details?id=com.google.android.apps.accessibility.auditor)

### CODE REVIEW

**Comment by mikehardy** [comment]
@dae you may actually be the authority here, I can get this in a 2.15.2 which we will do today but I'd love a set of :eyes: on it :pray:

**Comment by mikehardy** [comment]
I'm investigating unit test failures, they may be legit

**Comment by mikehardy** [comment]
```

com.ichi2.libanki.ModelTest > test_req FAILED
    java.lang.Exception: Main looper has queued unexecuted runnables. This might be the cause of the test failure. You might need a shadowOf(getMainLooper()).idle() call.
        at org.robolectric.android.internal.AndroidTestEnvironment.checkStateAfterTestFailure(AndroidTestEnvironment.java:502)
        at org.robolectric.RobolectricTestRunner$HelperTestRunner$1.evaluate(RobolectricTestRunner.java:581)
        at org.robolectric.internal.SandboxTestRunner$2.lambda$evaluate$0(SandboxTestRunner.java:278)
        at org.robolectric.internal.bytecode.Sandbox.lambda$runOnMainThread$0(Sandbox.java:89)
        at java.base/java.util.concurrent.FutureTask.run(FutureTask.java:264)
        at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128)
        at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628)
        at java.base/java.lang.Thread.run(Thread.java:829)

        Caused by:
        java.lang.AssertionError: expected:<[1,"all",[1,2]]> but was:<[1,"none",[]]>
            at org.junit.Assert.fail(Assert.java:89)
            at org.junit.Assert.failNotEquals(Assert.java:835)
            at org.junit.Assert.assertEquals(Assert.java:120)
            at org.junit.Assert.assertEquals(Assert.java:146)
            at com.ichi2.libanki.ModelTest.test_req(ModelTest.java:491)


com.ichi2.anki.CardTemplateEditorTest > testDeleteTemplateWithSelectivelyGeneratedCards FAILED
    java.lang.AssertionError: should be two cards now expected:<2> but was:<1>
        at org.junit.Assert.fail(Assert.java:89)
        at org.junit.Assert.failNotEquals(Assert.java:835)
        at org.junit.Assert.assertEquals(Assert.java:647)
        at com.ichi2.anki.CardTemplateEditorTest.testDeleteTemplateWithSelectivelyGeneratedCards(CardTemplateEditorTest.java:295)

com.ichi2.anki.CardTemplateEditorTest > testDeletePendingAddExistingCardCount FAILED
    java.lang.AssertionError: card generation should result in two cards expected:<2> but was:<1>
        at org.junit.Assert.fail(Assert.java:89)
        at org.junit.Assert.failNotEquals(Assert.java:835)
        at org.junit.Assert.assertEquals(Assert.java:647)
        at com.ichi2.anki.CardTemplateEditorTest.testDeletePendingAddExistingCardCount(CardTemplateEditorTest.java:468)
```

**Comment by mikehardy** [comment]
Fixed and re-pushed

Replacement type should not be altered in general
Other types should not just replaceFirst but the replacement should be anchored at start of string so that spaces in field name (like `#Add Reverse`) are never altered (only `# Add Reverse` should be trimmed)

**Comment by mikehardy** [comment]
Good to go pending review

File: AnkiDroid/src/main/java/com/ichi2/libanki/template/Tokenizer.java [hunk 1]
```diff
@@ -200,11 +200,11 @@ public boolean equals(@Nullable Object obj) {
         }
         switch (start.charAt(0)) {
             case '#':
-                return new Token(OPEN_CONDITIONAL, start.substring(1));
+                return new Token(OPEN_CONDITIONAL, start.substring(1).replaceFirst("^\\s+", ""));
```
Comment (Arthur-Milchior) [inline AnkiDroid/src/main/java/com/ichi2/libanki/template/Tokenizer.java:203]: Please use `trim` instead.
The end is already trimmed, so it will actually only trim the start.

This method runs once by card during "empty card" so it is really important for it to be optimized. [`trim`](https://android.googlesource.com/platform/libcore/+/0806909/libdvm/src/main/java/java/lang/String.java#1558) is really optimized while regexps are not

Comment (mikehardy) [reply AnkiDroid/src/main/java/com/ichi2/libanki/template/Tokenizer.java:203]: A great point - happy to change it, and trim expresses exactly what I want to do anyway, it's the exact special case of my regex

**Review by Arthur-Milchior — APPROVED** [review]

**Review by david-allison-1 — APPROVED** [review]
👍
